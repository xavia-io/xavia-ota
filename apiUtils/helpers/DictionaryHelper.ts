import { Dictionary } from 'structured-headers';

export class DictionaryHelper {
  static convertToDictionaryItemsRepresentation(obj: { [key: string]: string }): Dictionary {
    return new Map(
      Object.entries(obj).map(([k, v]) => {
        return [k, [v, new Map()]];
      })
    );
  }
}
