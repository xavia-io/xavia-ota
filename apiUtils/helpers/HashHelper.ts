import crypto, { BinaryToTextEncoding } from 'crypto';

export class HashHelper {
  static createHash(
    file: Buffer,
    hashingAlgorithm: string,
    encoding: BinaryToTextEncoding
  ): string {
    return crypto.createHash(hashingAlgorithm).update(file).digest(encoding);
  }

  static getBase64URLEncoding(base64EncodedString: string): string {
    return base64EncodedString.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
  }

  static convertSHA256HashToUUID(value: string): string {
    return `${value.slice(0, 8)}-${value.slice(8, 12)}-${value.slice(12, 16)}-${value.slice(
      16,
      20
    )}-${value.slice(20, 32)}`;
  }

  static signRSASHA256(data: string, privateKey: string): string {
    const sign = crypto.createSign('RSA-SHA256');
    sign.update(data, 'utf8');
    sign.end();
    return sign.sign(privateKey, 'base64');
  }
}
