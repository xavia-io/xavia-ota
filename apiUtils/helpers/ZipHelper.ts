import AdmZip from 'adm-zip';

import { StorageFactory } from '../storage/StorageFactory';

interface CachedZip {
  zip: AdmZip;
  timestamp: number;
}

export class ZipHelper {
  private static zipCache: Map<string, CachedZip> = new Map();
  private static readonly CACHE_DURATION = 5 * 60 * 1000; // 5 minutes

  static async getZipFromStorage(updateBundlePath: string): Promise<AdmZip> {
    const storage = StorageFactory.getStorage();
    const cached = this.zipCache.get(updateBundlePath);

    if (cached && Date.now() - cached.timestamp < this.CACHE_DURATION) {
      return cached.zip;
    }

    const zipBuffer = await storage.downloadFile(`${updateBundlePath}.zip`);
    const zip = new AdmZip(zipBuffer);
    this.zipCache.set(updateBundlePath, { zip, timestamp: Date.now() });
    return zip;
  }

  static async getFileFromZip(zip: AdmZip, filePath: string): Promise<Buffer> {
    const entries = zip.getEntries();
    const entry = entries.find((entry) => entry.entryName === filePath);
    if (!entry) throw new Error(`File not found in zip: ${filePath}`);
    return entry.getData();
  }
}
