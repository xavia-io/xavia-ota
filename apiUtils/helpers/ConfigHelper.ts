import { ZipHelper } from './ZipHelper';

export class ConfigHelper {
  static async getExpoConfigAsync({
    updateBundlePath,
    runtimeVersion,
  }: {
    updateBundlePath: string;
    runtimeVersion: string;
  }): Promise<any> {
    try {
      const zip = await ZipHelper.getZipFromStorage(updateBundlePath);
      const configBuffer = await ZipHelper.getFileFromZip(zip, 'expoconfig.json');
      return JSON.parse(configBuffer.toString('utf-8'));
    } catch (error) {
      throw new Error(
        `No expo config json found with runtime version: ${runtimeVersion}. Error: ${error}`
      );
    }
  }

  static getPrivateKey(): string | null {
    const privateKey = Buffer.from(process.env.PRIVATE_KEY_BASE_64 ?? '', 'base64').toString(
      'utf8'
    );
    return privateKey ?? null;
  }
}
