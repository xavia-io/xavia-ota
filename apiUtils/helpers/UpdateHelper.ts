import mime from 'mime';

import { HashHelper } from './HashHelper';
import { ZipHelper } from './ZipHelper';
import { StorageFactory } from '../storage/StorageFactory';

export class NoUpdateAvailableError extends Error {}
export type GetAssetMetadataArg =
  | {
      updateBundlePath: string;
      filePath: string;
      ext: null;
      isLaunchAsset: true;
      runtimeVersion: string;
      platform: string;
    }
  | {
      updateBundlePath: string;
      filePath: string;
      ext: string;
      isLaunchAsset: false;
      runtimeVersion: string;
      platform: string;
    };

export class UpdateHelper {
  static async getLatestUpdateBundlePathForRuntimeVersionAsync(
    runtimeVersion: string
  ): Promise<string> {
    const storage = StorageFactory.getStorage();
    const updatesDirectoryForRuntimeVersion = `updates/${runtimeVersion}`;

    if (!(await storage.fileExists(updatesDirectoryForRuntimeVersion))) {
      throw new Error('Unsupported runtime version');
    }

    const zipFiles = (await storage.listFiles(updatesDirectoryForRuntimeVersion))
      .filter((file) => file.name.endsWith('.zip'))
      .sort((a, b) => parseInt(b.name.split('.')[0], 10) - parseInt(a.name.split('.')[0], 10));

    if (!zipFiles.length) {
      throw new Error(`No updates found for runtime version: ${runtimeVersion}`);
    }

    return `${updatesDirectoryForRuntimeVersion}/${zipFiles[0].name.replace('.zip', '')}`;
  }

  static async getAssetMetadataAsync(arg: GetAssetMetadataArg) {
    const zip = await ZipHelper.getZipFromStorage(arg.updateBundlePath);
    const asset = await ZipHelper.getFileFromZip(zip, arg.filePath);

    const assetHash = HashHelper.getBase64URLEncoding(
      HashHelper.createHash(asset, 'sha256', 'base64')
    );
    const key = HashHelper.createHash(asset, 'md5', 'hex');
    const keyExtensionSuffix = arg.isLaunchAsset ? 'bundle' : arg.ext;
    const contentType = arg.isLaunchAsset ? 'application/javascript' : mime.getType(arg.ext);

    return {
      hash: assetHash,
      key,
      fileExtension: `.${keyExtensionSuffix}`,
      contentType,
      url: `${process.env.HOST}/api/assets?asset=${arg.filePath}&runtimeVersion=${arg.runtimeVersion}&platform=${arg.platform}`,
    };
  }

  static async getMetadataAsync({
    updateBundlePath,
    runtimeVersion,
  }: {
    updateBundlePath: string;
    runtimeVersion: string;
  }) {
    try {
      const zip = await ZipHelper.getZipFromStorage(updateBundlePath);
      const metadataBuffer = await ZipHelper.getFileFromZip(zip, 'metadata.json');
      const metadataJson = JSON.parse(metadataBuffer.toString('utf-8'));

      return {
        metadataJson,
        createdAt: new Date().toISOString(),
        id: HashHelper.createHash(metadataBuffer, 'sha256', 'hex'),
      };
    } catch (error) {
      throw new Error(`No metadata found with runtime version: ${runtimeVersion}. Error: ${error}`);
    }
  }

  static async createRollBackDirectiveAsync(updateBundlePath: string) {
    try {
      const zip = await ZipHelper.getZipFromStorage(updateBundlePath);
      const hasRollback = zip.getEntry('rollback') !== null;

      if (hasRollback) {
        return {
          type: 'rollBackToEmbedded',
          parameters: {
            commitTime: new Date().toISOString(),
          },
        };
      }
      throw new Error('No rollback file found');
    } catch (error) {
      throw new Error(`No rollback found. Error: ${error}`);
    }
  }

  static async createNoUpdateAvailableDirectiveAsync() {
    return {
      type: 'noUpdateAvailable',
    };
  }
}
