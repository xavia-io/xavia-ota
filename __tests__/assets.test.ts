import { createMocks } from 'node-mocks-http';

import { UpdateHelper } from '../apiUtils/helpers/UpdateHelper';
import { ZipHelper } from '../apiUtils/helpers/ZipHelper';
import assetsEndpoint from '../pages/api/assets';

jest.mock('../apiUtils/helpers/UpdateHelper');
jest.mock('../apiUtils/helpers/ZipHelper');

describe('Assets API', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should return 400 if asset path is missing', async () => {
    const { req, res } = createMocks({
      method: 'GET',
      query: {
        platform: 'ios',
        runtimeVersion: '1.0.0',
      },
    });

    await assetsEndpoint(req, res);
    expect(res._getStatusCode()).toBe(400);
  });

  it('should return 400 if platform is invalid', async () => {
    const { req, res } = createMocks({
      method: 'GET',
      query: {
        asset: 'test.png',
        platform: 'web',
        runtimeVersion: '1.0.0',
      },
    });

    await assetsEndpoint(req, res);
    expect(res._getStatusCode()).toBe(400);
  });

  it('should serve asset successfully', async () => {
    const mockMetadata = {
      metadataJson: {
        fileMetadata: {
          ios: {
            assets: [{ path: 'test.png', ext: '.png' }],
            bundle: 'bundle.js',
          },
        },
      },
    };

    (UpdateHelper.getLatestUpdateBundlePathForRuntimeVersionAsync as jest.Mock).mockResolvedValue(
      'path/to/update'
    );
    (UpdateHelper.getMetadataAsync as jest.Mock).mockResolvedValue(mockMetadata);
    (ZipHelper.getZipFromStorage as jest.Mock).mockResolvedValue({});
    (ZipHelper.getFileFromZip as jest.Mock).mockResolvedValue(Buffer.from('test'));

    const { req, res } = createMocks({
      method: 'GET',
      query: {
        asset: 'test.png',
        platform: 'ios',
        runtimeVersion: '1.0.0',
      },
    });

    await assetsEndpoint(req, res);
    expect(res._getStatusCode()).toBe(200);
    expect(res._getData()).toMatchSnapshot();
  });
});
