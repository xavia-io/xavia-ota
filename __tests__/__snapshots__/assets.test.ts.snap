// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assets API should serve asset successfully 1`] = `""`;
